xenomint/state.go

```diff
@@ -0,0 +1,494 @@
+		id             uint64
+		ierr           error
+		cnames, ctypes []string
+		data           [][]interface{}
+	)
+	if tx, ierr = s.strg.DirtyReader().Begin(); ierr != nil {
+		err = errors.Wrap(ierr, "open tx failed")
+		return
+	}
+	defer tx.Rollback()
+	// FIXME(leventeliu): lock free but not consistent.
+	id = s.getID()
+	for i, v := range req.Payload.Queries {
+		if cnames, ctypes, data, ierr = readSingle(tx, &v); ierr != nil {
+			err = errors.Wrapf(ierr, "query at #%d failed", i)
+			return
+		}
+	}
+	// Build query response
+	ref = &QueryTracker{req: req}
+	resp = &types.Response{
+		Header: types.SignedResponseHeader{
+			ResponseHeader: types.ResponseHeader{
+				Request:   req.Header,
+				NodeID:    "",
+				Timestamp: time.Now(),
+				RowCount:  uint64(len(data)),
+				LogOffset: id,
+			},
+		},
+		Payload: types.ResponsePayload{
+			Columns:   cnames,
+			DeclTypes: ctypes,
+			Rows:      buildRowsFromNativeData(data),
+		},
+	}
+	return
+}
+
+func (s *State) writeSingle(q *types.Query) (res sql.Result, err error) {
+	if res, err = s.unc.Exec(q.Pattern, buildArgsFromSQLNamedArgs(q.Args)...); err == nil {
+		s.incSeq()
+	}
+	return
+}
+
+func (s *State) setSavepoint() (savepoint uint64) {
+	savepoint = s.getID()
+	s.unc.Exec("SAVEPOINT ?", savepoint)
+	return
+}
+
+func (s *State) rollbackTo(savepoint uint64) {
+	s.rollbackID(savepoint)
+	s.unc.Exec("ROLLBACK TO ?", savepoint)
+}
+
+func (s *State) write(req *types.Request) (ref *QueryTracker, resp *types.Response, err error) {
+	var (
+		ierr      error
+		savepoint uint64
+		query     = &QueryTracker{req: req}
+	)
+	// TODO(leventeliu): savepoint is a sqlite-specified solution for nested transaction.
+	func() {
+		s.Lock()
+		defer s.Unlock()
+		savepoint = s.getID()
+		for i, v := range req.Payload.Queries {
+			if _, ierr = s.writeSingle(&v); ierr != nil {
+				err = errors.Wrapf(ierr, "execute at #%d failed", i)
+				s.rollbackTo(savepoint)
+				return
+			}
+		}
+		s.setSavepoint()
+		s.pool.enqueue(savepoint, query)
+	}()
+	// Build query response
+	ref = query
+	resp = &types.Response{
+		Header: types.SignedResponseHeader{
+			ResponseHeader: types.ResponseHeader{
+				Request:   req.Header,
+				NodeID:    "",
+				Timestamp: time.Now(),
+				RowCount:  0,
+				LogOffset: savepoint,
+			},
+		},
+	}
+	return
+}
+
+func (s *State) replay(req *types.Request, resp *types.Response) (err error) {
+	var (
+		ierr      error
+		savepoint uint64
+		query     = &QueryTracker{req: req, resp: resp}
+	)
+	s.Lock()
+	defer s.Unlock()
+	savepoint = s.getID()
+	if resp.Header.ResponseHeader.LogOffset != savepoint {
+		err = ErrQueryConflict
+		return
+	}
+	for i, v := range req.Payload.Queries {
+		if _, ierr = s.writeSingle(&v); ierr != nil {
+			err = errors.Wrapf(ierr, "execute at #%d failed", i)
+			s.rollbackTo(savepoint)
+			return
+		}
+	}
+	s.setSavepoint()
+	s.pool.enqueue(savepoint, query)
+	return
+}
+
+func (s *State) ReplayBlock(block *types.Block) (err error) {
```

Comment:
**[golint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
exported method State.ReplayBlock should have comment or be unexported